File: jpp/l3/rust/src/main.rs
mod utils;
mod rsa;
mod dhsetup;
mod user;
mod ring;

fn print_result(alice_pub_key: u64, bob_pub_key: u64, message: u64, cipher: u64, decrypted: u64) {
    println!("Alice's public key: {}", alice_pub_key);
    println!("Bob's public key: {}", bob_pub_key);
    println!("Message: {}", message);
    println!("Cipher: {}", cipher);
    println!("Decrypted: {}", decrypted);
    println!("Decryption successful: {}", decrypted == message);
}

fn rsa_example() {
    println!("RSA encryption example:\n");

    let alice = rsa::RSA::<10007, 10009>::new();
    let bob = rsa::RSA::<10007, 10009>::new();

    let message: u64 = 12345;
    let cipher = alice.encrypt(message, bob.public_key());
    let decrypted = bob.decrypt(cipher);

    print_result(alice.public_key(), bob.public_key(), message, cipher, decrypted);
}

fn diff_hell() {
    println!("Diffie-Hellman key exchange example:\n");

    let dh = dhsetup::DHSetup::<1234567891>::new();
    let mut alice = user::User::<1234567891>::new(&dh);
    let mut bob = user::User::<1234567891>::new(&dh);

    println!("Generator: {}", dh.get_generator());

    alice.set_key(bob.get_public_key());
    bob.set_key(alice.get_public_key());
    let message: u64 = 12345;
    let cipher = alice.encrypt(message);
    let decrypted = bob.decrypt(cipher);
    print_result(alice.get_public_key(), bob.get_public_key(), message, cipher, decrypted);
}

fn main() {
    rsa_example();
    println!();
    diff_hell();
}

File: jpp/l3/cpp/main.cpp
#include <iostream>
#include <format>

#include "rsa.hpp"
#include "diff-hell.hpp"

template<typename T>
void print_results(T& alice, T& bob, std::size_t cipher, std::size_t decrypted, std::size_t message) {
    std::cout << std::format("Bob's public key: {}\n", bob.getPublicKey());
    std::cout << std::format("Alice's public key: {}\n", alice.getPublicKey());
    std::cout << std::format("Message: {}\n", message);
    std::cout << std::format("Cipher: {}\n", cipher);
    std::cout << std::format("Decrypted: {}\n", decrypted);
    std::cout << std::format("Decryption successful: {}\n", decrypted == message ? "Yes" : "No");
}

void rsa_example() {
    std::cout << "RSA encryption example:\n";

    constexpr auto P = 10007, Q = 10009;
    RSA<P, Q> bob;
    RSA<P, Q> alice;

    std::size_t message = 12345;
    auto cipher = bob.encrypt(message, alice.getPublicKey());
    auto decrypted = alice.decrypt(cipher);

    print_results(alice, bob, cipher, decrypted, message);
}

void diff_hell() {
    std::cout << "Diffie-Hellman key exchange example:\n";

    constexpr auto P = 1234567891;
    DHSetup<P> setup;
    
    User<P> alice(setup);
    User<P> bob(setup);

    alice.setKey(bob.getPublicKey());
    bob.setKey(alice.getPublicKey());

    std::size_t message = 12345;
    auto cipher = alice.encrypt(message);
    auto decrypted = bob.decrypt(cipher);

    std::cout << std::format("Generator: {}\n", setup.getGenerator());
    print_results(alice, bob, cipher, decrypted, message);
}

int main() {
    rsa_example();
    std::cout << "\n\n";
    diff_hell();
    return 0;
}

